// react-simple-animate/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};
    var ownKeys = Object.keys(source);

    if (typeof Object.getOwnPropertySymbols === 'function') {
      ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
        return Object.getOwnPropertyDescriptor(source, sym).enumerable;
      }));
    }

    ownKeys.forEach(function (key) {
      _defineProperty(target, key, source[key]);
    });
  }

  return target;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest();
}

function _toConsumableArray(arr) {
  return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread();
}

function _arrayWithoutHoles(arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) arr2[i] = arr[i];

    return arr2;
  }
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArray(iter) {
  if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter);
}

function _iterableToArrayLimit(arr, i) {
  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance");
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance");
}

// 
var mapSequenceOverProps = (function (props, id) {
  var animationStates = props.animationStates,
      restProps = _objectWithoutProperties(props, ["animationStates"]);

  if (!id || !animationStates || !animationStates[id]) return props;

  var stateCopy = _objectSpread({}, animationStates[id]);

  return _objectSpread({}, stateCopy, restProps);
});

function attributesGenerator(props) {
  var willComplete = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var isMountWithPlay = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
  var animationStates = props.animationStates,
      sequenceId = props.sequenceId,
      sequenceIndex = props.sequenceIndex;
  var id = sequenceId || sequenceIndex;

  var _mapSequenceOverProps = mapSequenceOverProps(props, id),
      play = _mapSequenceOverProps.play,
      startStyle = _mapSequenceOverProps.startStyle,
      endStyle = _mapSequenceOverProps.endStyle,
      onCompleteStyle = _mapSequenceOverProps.onCompleteStyle,
      _mapSequenceOverProps2 = _mapSequenceOverProps.durationSeconds,
      durationSeconds = _mapSequenceOverProps2 === void 0 ? 0.3 : _mapSequenceOverProps2,
      _mapSequenceOverProps3 = _mapSequenceOverProps.delaySeconds,
      delaySeconds = _mapSequenceOverProps3 === void 0 ? 0 : _mapSequenceOverProps3,
      _mapSequenceOverProps4 = _mapSequenceOverProps.easeType,
      easeType = _mapSequenceOverProps4 === void 0 ? 'linear' : _mapSequenceOverProps4,
      className = _mapSequenceOverProps.className,
      _mapSequenceOverProps5 = _mapSequenceOverProps.reverseDurationSeconds,
      reverseDurationSeconds = _mapSequenceOverProps5 === void 0 ? 0 : _mapSequenceOverProps5,
      _mapSequenceOverProps6 = _mapSequenceOverProps.reverseDelaySeconds,
      reverseDelaySeconds = _mapSequenceOverProps6 === void 0 ? 0 : _mapSequenceOverProps6,
      forwardedRef = _mapSequenceOverProps.forwardedRef;

  var style = startStyle;
  var transition = "all ".concat(durationSeconds, "s ").concat(easeType, " ").concat(delaySeconds, "s");

  if (play !== undefined && !play && (reverseDurationSeconds || reverseDelaySeconds)) {
    transition = "all ".concat(reverseDurationSeconds || durationSeconds, "s ").concat(easeType, " ").concat(reverseDelaySeconds, "s");
  } else if (!isMountWithPlay) {
    if (willComplete && onCompleteStyle && play) {
      style = onCompleteStyle;
      transition = '';
    } else if (play || (id || id === 0) && animationStates && Object.keys(animationStates).length && animationStates[id] && animationStates[id].play) {
      style = endStyle;
    }
  }

  return {
    className: className,
    style: _objectSpread({}, style, {
      transition: transition
    }),
    ref: forwardedRef
  };
}

// 
var msToSec = (function (ms) {
  return (parseFloat(ms) || 0) * 1000;
});

// 
var calculateTotalDuration = (function (_ref) {
  var _ref$durationSeconds = _ref.durationSeconds,
      durationSeconds = _ref$durationSeconds === void 0 ? 0 : _ref$durationSeconds,
      _ref$delaySeconds = _ref.delaySeconds,
      delaySeconds = _ref$delaySeconds === void 0 ? 0 : _ref$delaySeconds,
      _ref$reverseDurationS = _ref.reverseDurationSeconds,
      reverseDurationSeconds = _ref$reverseDurationS === void 0 ? 0 : _ref$reverseDurationS,
      play = _ref.play;
  var duration = parseFloat(play ? durationSeconds : reverseDurationSeconds || durationSeconds) + parseFloat(delaySeconds);
  return msToSec(duration);
});

var AnimateContext = React__default.createContext({
  animationStates: {},
  register: function register() {}
});

var AnimateGroup =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(AnimateGroup, _React$PureComponent);

  function AnimateGroup() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, AnimateGroup);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(AnimateGroup)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      animationStates: {}
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "timers", {});

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "animations", {});

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "clearTimers", function () {
      return Object.values(_this.timers).forEach(function (timer) {
        return clearTimeout(timer);
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "setupAnimationTimers", function (_ref) {
      var totalDuration = _ref.totalDuration,
          id = _ref.id,
          restAttributes = _ref.restAttributes,
          play = _ref.play;
      var duration = totalDuration - (restAttributes.overlaySeconds || 0) * 1000;
      _this.timers[id] = setTimeout(function () {
        _this.setState(function (previousState) {
          var stateCopy = _objectSpread({}, previousState.animationStates);

          if (!stateCopy[id]) stateCopy[id] = {};
          Object.entries(restAttributes).forEach(function (_ref2) {
            var _ref3 = _slicedToArray(_ref2, 2),
                key = _ref3[0],
                value = _ref3[1];

            stateCopy[id][key] = value;
          });
          stateCopy[id].play = play;
          return {
            animationStates: stateCopy
          };
        });
      }, duration < 0 ? 0 : duration);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "calculateSequences", function () {
      var _this$props = _this.props,
          sequences = _this$props.sequences,
          reverseSequences = _this$props.reverseSequences,
          play = _this$props.play;
      var sequencesToAnimate = Array.isArray(sequences) && sequences.length ? sequences : Object.values(_this.animations);
      var reverseSequencesToAnimate = Array.isArray(reverseSequences) && reverseSequences.length ? reverseSequences : _toConsumableArray(sequencesToAnimate).reverse();
      return (play ? sequencesToAnimate : reverseSequencesToAnimate).reduce(function (previous, current, currentIndex) {
        var sequenceId = current.sequenceId,
            sequenceIndex = current.sequenceIndex,
            restAttributes = _objectWithoutProperties(current, ["sequenceId", "sequenceIndex"]);

        var id = sequenceId === undefined && sequenceIndex === undefined ? currentIndex : sequenceId || sequenceIndex;
        var totalDuration = previous + calculateTotalDuration(_objectSpread({}, _this.animations[id], {
          play: play,
          restAttributes: restAttributes
        }));

        _this.setupAnimationTimers({
          id: id,
          totalDuration: currentIndex === 0 ? 0 : previous,
          restAttributes: restAttributes,
          play: play
        });

        return totalDuration;
      }, 0);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "register", function (props) {
      var sequenceIndex = props.sequenceIndex,
          sequenceId = props.sequenceId;
      var id = sequenceId || sequenceIndex;
      if (id === undefined || sequenceIndex && sequenceIndex < 0 || sequenceId && sequenceId === '') return;
      _this.animations[id] = _objectSpread({}, props);
    });

    return _this;
  }

  _createClass(AnimateGroup, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.props.play && this.calculateSequences();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.play !== prevProps.play) {
        this.clearTimers();
        this.calculateSequences();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.clearTimers();
    }
  }, {
    key: "render",
    value: function render() {
      return React__default.createElement(AnimateContext.Provider, {
        value: {
          animationStates: this.state.animationStates,
          register: this.register
        }
      }, this.props.children);
    }
  }]);

  return AnimateGroup;
}(React__default.PureComponent);

_defineProperty(AnimateGroup, "displayName", 'AnimateGroup');

_defineProperty(AnimateGroup, "defaultProps", {
  sequences: [],
  reverseSequences: []
});

var AnimateChild =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(AnimateChild, _React$PureComponent);

  function AnimateChild(props) {
    var _this;

    _classCallCheck(this, AnimateChild);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(AnimateChild).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "isMountWithPlay", false);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      willComplete: false,
      play: false
    });

    var play = props.play,
        register = props.register;
    register && register(_this.props);

    if (play) {
      _this.isMountWithPlay = true;
      _this.initialPlayTimer = setTimeout(function () {
        _this.isMountWithPlay = false;

        _this.forceUpdate();
      }, msToSec(props.delaySeconds));
    }

    return _this;
  }

  _createClass(AnimateChild, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      var _this2 = this;

      var _this$props = this.props,
          delaySeconds = _this$props.delaySeconds,
          play = _this$props.play,
          onCompleteStyle = _this$props.onCompleteStyle,
          durationSeconds = _this$props.durationSeconds,
          onComplete = _this$props.onComplete,
          animationStates = _this$props.animationStates,
          sequenceId = _this$props.sequenceId,
          sequenceIndex = _this$props.sequenceIndex;
      var id = sequenceId || sequenceIndex;

      if ((onComplete || onCompleteStyle) && !this.state.willComplete && (play || id && animationStates && Object.keys(animationStates).length && animationStates[id] && animationStates[id].play)) {
        clearTimeout(this.completeTimeout);
        this.completeTimeout = setTimeout(function () {
          _this2.setState({
            willComplete: true
          });

          onComplete && onComplete();
        }, msToSec(parseFloat(delaySeconds) + parseFloat(durationSeconds)));
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.completeTimeout);
      clearTimeout(this.unMountTimeout);
      clearTimeout(this.initialPlayTimer);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          _this$props2$tag = _this$props2.tag,
          tag = _this$props2$tag === void 0 ? 'div' : _this$props2$tag,
          children = _this$props2.children,
          render = _this$props2.render;
      var props = attributesGenerator(this.props, this.state.willComplete, this.isMountWithPlay);
      return render ? render(props) : React__default.createElement(tag, props, children);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var animationStates = nextProps.animationStates,
          play = nextProps.play,
          sequenceId = nextProps.sequenceId,
          sequenceIndex = nextProps.sequenceIndex,
          onCompleteStyle = nextProps.onCompleteStyle,
          register = nextProps.register;
      var id = sequenceId || sequenceIndex;
      var currentPlay = play;
      register && register(nextProps);

      if (id !== undefined && animationStates && animationStates[id]) {
        var state = animationStates[id];
        currentPlay = state.play;
      }

      return _objectSpread({}, onCompleteStyle && prevState.willComplete ? {
        willComplete: !(play && !prevState.play && prevState.willComplete)
      } : null, currentPlay !== prevState.play ? {
        play: currentPlay
      } : null);
    }
  }]);

  return AnimateChild;
}(React__default.PureComponent); // $FlowIgnoreLine: flow complain about React.forwardRef disable for now

_defineProperty(AnimateChild, "displayName", 'Animate');

_defineProperty(AnimateChild, "defaultProps", {
  durationSeconds: 0.3,
  delaySeconds: 0,
  easeType: 'linear',
  sequenceId: undefined,
  sequenceIndex: undefined,
  animationStates: undefined
});

var animate = React__default.forwardRef(function (props, ref) {
  return React__default.createElement(AnimateContext.Consumer, null, function (_ref) {
    var _ref$animationStates = _ref.animationStates,
        animationStates = _ref$animationStates === void 0 ? {} : _ref$animationStates,
        _ref$register = _ref.register,
        register = _ref$register === void 0 ? undefined : _ref$register;
    return React__default.createElement(AnimateChild, _extends({}, _objectSpread({}, props, {
      animationStates: animationStates,
      register: register
    }), {
      forwardedRef: ref
    }));
  });
});

// 
function createStyle(_ref) {
  var keyframes = _ref.keyframes,
      animationName = _ref.animationName;
  var animationLength = keyframes.length;
  return "".concat(keyframes.reduce(function (previous, keyframe, currentIndex) {
    if (typeof keyframe === 'string') {
      return "".concat(previous, " ").concat(animationLength === 2 ? currentIndex * 100 : parseFloat((100 / (animationLength - 1)).toFixed(2)) * currentIndex, "% {").concat(keyframe, "}");
    } // $FlowIgnoreLine:


    return "".concat(previous, " ").concat(Object.keys(keyframe)[0], "% {").concat(Object.values(keyframe)[0], "}");
  }, "@keyframes ".concat(animationName, " {")), "}");
}

// 
function createTag(_ref) {
  var keyframes = _ref.keyframes,
      animationName = _ref.animationName;
  var styleTag = document.querySelector('style[data-id=rsi]');

  if (!styleTag) {
    styleTag = document.createElement('style');
    styleTag.setAttribute('data-id', 'rsi'); // $FlowIgnoreLine:

    document.head.appendChild(styleTag);
  } // $FlowIgnoreLine


  var index = styleTag.sheet.cssRules.length || 0;

  try {
    // $FlowIgnoreLine
    styleTag.sheet.insertRule(createStyle({
      keyframes: keyframes,
      animationName: animationName
    }), index);
  } catch (e) {
    console.error('react simple animate, error found during insert style ', e); // eslint-disable-line no-console
  }

  return {
    styleTag: styleTag,
    index: index
  };
}

// 
function createRandomName() {
  return "RSI-".concat(Math.random().toString(36).substr(2, 9));
}

// 
var deleteRules = (function (sheet, deleteName) {
  var index = Object.values(sheet.cssRules).findIndex(function (_ref) {
    var name = _ref.name;
    return name === deleteName;
  });

  if (index >= 0) {
    sheet.deleteRule(index);
  }
});

var AnimateKeyframesChild =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(AnimateKeyframesChild, _React$PureComponent);

  function AnimateKeyframesChild() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, AnimateKeyframesChild);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(AnimateKeyframesChild)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      play: false
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "createStyleAndTag", function () {
      var keyframes = _this.props.keyframes;
      _this.animationName = createRandomName();

      var _createTag = createTag({
        animationName: _this.animationName,
        keyframes: keyframes
      }),
          styleTag = _createTag.styleTag,
          index = _createTag.index;

      _this.styleTag = styleTag;
      _this.index = index;
    });

    return _this;
  }

  _createClass(AnimateKeyframesChild, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props = this.props,
          register = _this$props.register,
          play = _this$props.play;
      this.createStyleAndTag();
      register && register(this.props);

      if (play) {
        this.forceUpdate();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      deleteRules(this.styleTag.sheet, this.animationName);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          children = _this$props2.children,
          play = _this$props2.play,
          render = _this$props2.render,
          _this$props2$duration = _this$props2.durationSeconds,
          durationSeconds = _this$props2$duration === void 0 ? 0.3 : _this$props2$duration,
          _this$props2$delaySec = _this$props2.delaySeconds,
          delaySeconds = _this$props2$delaySec === void 0 ? 0 : _this$props2$delaySec,
          _this$props2$easeType = _this$props2.easeType,
          easeType = _this$props2$easeType === void 0 ? 'linear' : _this$props2$easeType,
          _this$props2$playStat = _this$props2.playState,
          playState = _this$props2$playStat === void 0 ? 'running' : _this$props2$playStat,
          _this$props2$directio = _this$props2.direction,
          direction = _this$props2$directio === void 0 ? 'normal' : _this$props2$directio,
          _this$props2$fillMode = _this$props2.fillMode,
          fillMode = _this$props2$fillMode === void 0 ? 'none' : _this$props2$fillMode,
          _this$props2$iteratio = _this$props2.iterationCount,
          iterationCount = _this$props2$iteratio === void 0 ? 1 : _this$props2$iteratio;
      var style = play || this.state.play ? {
        animation: "".concat(durationSeconds, "s ").concat(easeType, " ").concat(delaySeconds, "s ").concat(iterationCount, " ").concat(direction, " ").concat(fillMode, " ").concat(playState, " ").concat(this.animationName)
      } : null;
      return render ? render({
        style: style
      }) : React__default.createElement("div", style ? {
        style: style
      } : null, children);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var animationStates = nextProps.animationStates,
          play = nextProps.play,
          sequenceId = nextProps.sequenceId,
          sequenceIndex = nextProps.sequenceIndex;
      var id = sequenceId || sequenceIndex;
      var currentPlay = play;

      if (id !== undefined && animationStates && animationStates[id]) {
        var state = animationStates[id];
        currentPlay = state.play;
      }

      return _objectSpread({}, currentPlay !== prevState.play ? {
        play: currentPlay
      } : null);
    }
  }]);

  return AnimateKeyframesChild;
}(React__default.PureComponent); // $FlowIgnoreLine: flow complain about React.forwardRef disable for now

_defineProperty(AnimateKeyframesChild, "displayName", 'AnimateKeyframes');

_defineProperty(AnimateKeyframesChild, "defaultProps", {
  durationSeconds: 0.3,
  delaySeconds: 0,
  easeType: 'linear',
  render: undefined,
  playState: 'running',
  direction: 'normal',
  fillMode: 'none',
  iterationCount: 1,
  children: undefined,
  sequenceId: undefined,
  sequenceIndex: undefined,
  register: undefined
});

var animateKeyframes = React__default.forwardRef(function (props, ref) {
  return React__default.createElement(AnimateContext.Consumer, null, function (_ref) {
    var _ref$animationStates = _ref.animationStates,
        animationStates = _ref$animationStates === void 0 ? {} : _ref$animationStates,
        _ref$register = _ref.register,
        register = _ref$register === void 0 ? undefined : _ref$register;
    return React__default.createElement(AnimateKeyframesChild, _extends({}, _objectSpread({}, props, {
      animationStates: animationStates,
      register: register
    }), {
      forwardedRef: ref
    }));
  });
});

function useAnimate() {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
    durationSeconds: 0.3,
    delaySeconds: 0,
    easeType: 'linear',
    play: false,
    endStyle: {}
  };
  var completeTimeout;
  var initialPlayTimer;
  var onComplete = props.onComplete,
      onCompleteStyle = props.onCompleteStyle,
      _props$delaySeconds = props.delaySeconds,
      delaySeconds = _props$delaySeconds === void 0 ? 0 : _props$delaySeconds,
      _props$durationSecond = props.durationSeconds,
      durationSeconds = _props$durationSecond === void 0 ? 0.3 : _props$durationSecond;

  var _useState = React.useState(props),
      _useState2 = _slicedToArray(_useState, 2),
      animateProps = _useState2[0],
      setPlay = _useState2[1];

  var play = animateProps.play,
      willComplete = animateProps.willComplete,
      isMountWithPlay = animateProps.isMountWithPlay;

  var playFunction = function playFunction(playValue) {
    var isMountWithPlayValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : animateProps.isMountWithPlay;
    setPlay(_objectSpread({}, props, {
      play: playValue,
      isMountWithPlay: isMountWithPlayValue
    }));
  };

  React.useEffect(function () {
    if ((onComplete || onCompleteStyle) && play) {
      clearTimeout(completeTimeout);
      completeTimeout = setTimeout(function () {
        setPlay(_objectSpread({}, animateProps, {
          willComplete: true,
          isMountWithPlay: false
        }));
        onComplete && onComplete();
      }, msToSec(parseFloat(delaySeconds) + parseFloat(durationSeconds)));
    }

    if (play && props.play && isMountWithPlay === undefined) {
      initialPlayTimer = setTimeout(function () {
        playFunction(play, false);
      }, msToSec(delaySeconds));
    }

    return function () {
      clearTimeout(completeTimeout);
      clearTimeout(initialPlayTimer);
    };
  }, [play]);
  return [{
    style: attributesGenerator(_objectSpread({}, props, {
      play: play
    }), willComplete, isMountWithPlay === undefined ? props.play : isMountWithPlay).style,
    play: play
  }, playFunction];
}

function useAnimateKeyframes(props) {
  var _props$durationSecond = props.durationSeconds,
      durationSeconds = _props$durationSecond === void 0 ? 0.3 : _props$durationSecond,
      _props$delaySeconds = props.delaySeconds,
      delaySeconds = _props$delaySeconds === void 0 ? 0 : _props$delaySeconds,
      _props$easeType = props.easeType,
      easeType = _props$easeType === void 0 ? 'linear' : _props$easeType,
      _props$direction = props.direction,
      direction = _props$direction === void 0 ? 'normal' : _props$direction,
      _props$fillMode = props.fillMode,
      fillMode = _props$fillMode === void 0 ? 'none' : _props$fillMode,
      _props$iterationCount = props.iterationCount,
      iterationCount = _props$iterationCount === void 0 ? 1 : _props$iterationCount,
      _props$playState = props.playState,
      playState = _props$playState === void 0 ? 'running' : _props$playState,
      keyframes = props.keyframes;

  var _useState = React.useState(props),
      _useState2 = _slicedToArray(_useState, 2),
      animateProps = _useState2[0],
      setPlay = _useState2[1];

  var playFunction = function playFunction(playValue) {
    setPlay(_objectSpread({}, props, {
      play: playValue
    }));
  };

  var play = animateProps.play,
      animationName = animateProps.animationName;
  React.useEffect(function () {
    var name = createRandomName();

    var _createTag = createTag({
      animationName: name,
      keyframes: keyframes
    }),
        styleTag = _createTag.styleTag;

    var localStyleTag = styleTag;
    setPlay(_objectSpread({}, props, {
      animationName: name
    }));
    return function () {
      deleteRules(localStyleTag.sheet, name);
    };
  }, []);
  var style = play ? {
    animation: "".concat(durationSeconds, "s ").concat(easeType, " ").concat(delaySeconds, "s ").concat(iterationCount, " ").concat(direction, " ").concat(fillMode, " ").concat(playState, " ").concat(animationName)
  } : null;
  return [{
    style: style,
    play: play,
    animationName: animationName
  }, function (playValue) {
    return playFunction(playValue);
  }];
}

function useAnimateGroup(props) {
  var nextDelaySeconds = 0;

  var _useState = React.useState(props),
      _useState2 = _slicedToArray(_useState, 2),
      _useState2$ = _useState2[0],
      sequences = _useState2$.sequences,
      reverseSequences = _useState2$.reverseSequences,
      play = _useState2$.play,
      animationNames = _useState2$.animationNames,
      setPlay = _useState2[1];

  var playFunction = function playFunction(playValue) {
    setPlay({
      sequences: sequences,
      reverseSequences: reverseSequences,
      play: playValue,
      animationNames: animationNames
    });
  };

  React.useEffect(function () {
    var localStyleTag;
    var localAnimationNames = [];
    sequences.forEach(function (_ref, i) {
      var _ref$keyframes = _ref.keyframes,
          keyframes = _ref$keyframes === void 0 ? false : _ref$keyframes;

      if (keyframes) {
        var animationName = createRandomName();
        localAnimationNames[i] = animationName;

        var _createTag = createTag({
          animationName: animationName,
          keyframes: keyframes
        }),
            styleTag = _createTag.styleTag;

        localStyleTag = styleTag;
      }
    });
    setPlay({
      sequences: sequences,
      play: play,
      reverseSequences: reverseSequences,
      animationNames: localAnimationNames
    });
    return function () {
      if (localStyleTag) {
        localAnimationNames.forEach(function (name) {
          // $FlowIgnoreLine
          deleteRules(localStyleTag.sheet, name);
        });
      }

      localAnimationNames = [];
    };
  }, []);
  var localSequences = (play ? sequences : reverseSequences) || [];
  var styles = (reverseSequences ? localSequences : sequences).map(function (prop, i) {
    var _prop$durationSeconds = prop.durationSeconds,
        durationSeconds = _prop$durationSeconds === void 0 ? 0.3 : _prop$durationSeconds,
        _prop$keyframes = prop.keyframes,
        keyframes = _prop$keyframes === void 0 ? false : _prop$keyframes,
        _prop$easeType = prop.easeType,
        easeType = _prop$easeType === void 0 ? 'linear' : _prop$easeType,
        _prop$delaySeconds = prop.delaySeconds,
        delaySeconds = _prop$delaySeconds === void 0 ? 0 : _prop$delaySeconds,
        _prop$iterationCount = prop.iterationCount,
        iterationCount = _prop$iterationCount === void 0 ? 1 : _prop$iterationCount,
        _prop$direction = prop.direction,
        direction = _prop$direction === void 0 ? 'normal' : _prop$direction,
        _prop$fillMode = prop.fillMode,
        fillMode = _prop$fillMode === void 0 ? 'none' : _prop$fillMode,
        _prop$playState = prop.playState,
        stylePlayState = _prop$playState === void 0 ? 'running' : _prop$playState,
        _prop$overlaySeconds = prop.overlaySeconds,
        overlaySeconds = _prop$overlaySeconds === void 0 ? 0 : _prop$overlaySeconds;
    nextDelaySeconds = durationSeconds + delaySeconds - overlaySeconds;
    nextDelaySeconds = nextDelaySeconds < 0 ? 0 : nextDelaySeconds;

    if (keyframes) {
      return play ? {
        styles: "".concat(durationSeconds, "s ").concat(easeType, " ").concat(i === 0 ? delaySeconds : nextDelaySeconds + delaySeconds, "s ").concat(iterationCount, " ").concat(direction, " ").concat(fillMode, " ").concat(stylePlayState, " ").concat(animationNames[i])
      } : null;
    }

    return attributesGenerator(_objectSpread({}, _objectSpread({}, prop, {
      delaySeconds: i === 0 ? delaySeconds : nextDelaySeconds + delaySeconds
    }), {
      play: play
    })).style;
  });
  return [{
    styles: styles,
    animationNames: animationNames,
    play: play
  }, playFunction];
}

//

exports.Animate = animate;
exports.AnimateGroup = AnimateGroup;
exports.AnimateKeyframes = animateKeyframes;
exports.useAnimate = useAnimate;
exports.useAnimateKeyframes = useAnimateKeyframes;
exports.useAnimateGroup = useAnimateGroup;
